fdbclient/FileBackupAgent.actor.cpp

```diff
@@ -458,8 +475,417 @@ Value makePadding(int size) {
 	return pad.substr(0, size);
 }
 
+struct IRangeFileWriter {
+public:
+	virtual Future<Void> padEnd(bool final) = 0;
+
+	virtual Future<Void> writeKV(Key k, Value v) = 0;
+
+	virtual Future<Void> writeKey(Key k) = 0;
+
+	virtual Future<Void> finish() = 0;
+
+	virtual ~IRangeFileWriter() {}
+};
+
+struct SnapshotFileBackupEncryptionKeys {
+	Reference<BlobCipherKey> textCipherKey;
+	Reference<BlobCipherKey> headerCipherKey;
+	StringRef ivRef;
+};
+
+// File Format handlers.
+// Both Range and Log formats are designed to be readable starting at any BACKUP_RANGEFILE_BLOCK_SIZE boundary
+// so they can be read in parallel.
+//
+// Writer instances must be kept alive while any member actors are in progress.
+//
+// EncryptedRangeFileWriter must be used as follows:
+//   1 - writeKey(key) the queried key range begin
+//   2 - writeKV(k, v) each kv pair to restore
+//   3 - writeKey(key) the queried key range end
+//   4 - finish()
+//
+// EncryptedRangeFileWriter will insert the required padding, header, and extra
+// end/begin keys around the 1MB boundaries as needed.
+//
+// Example:
+//   The range a-z is queries and returns c-j which covers 3 blocks across 2 tenants.
+//   The client code writes keys in this sequence:
+//             t1a t1c t1d t1e t1f t1g t2h t2i t2j t2z
+//
+//   H = header   P = padding   a...z = keys  v = value | = block boundary
+//
+//   Encoded file:  H t1a t1cv t1dv t1ev P | H t1e t1ev t1fv t1gv t2 P | H t2 t2hv t2iv t2jv t2z
+//   Decoded in blocks yields:
+//           Block 1: range [t1a, t1e) with kv pairs t1cv, t1dv
+//           Block 2: range [t1e, t2) with kv pairs t1ev, t1fv, t1gv
+//           Block 3: range [t2, t2z) with kv pairs t2hv, t2iv, t2jv
+//
+//   NOTE: All blocks except for the final block will have one last
+//   value which will not be used.  This isn't actually a waste since
+//   if the next KV pair wouldn't fit within the block after the value
+//   then the space after the final key to the next 1MB boundary would
+//   just be padding anyway.
+//
+//   NOTE: For the EncryptedRangeFileWriter blocks will be split either on the BACKUP_RANGEFILE_BLOCK_SIZE boundary or
+//   when a new tenant id is encountered. If a block is split for crossing tenant boundaries then the last key will be
+//   truncated to just the tenant prefix and the value will be empty (to avoid having sensitive data of one tenant be
+//   encrypted with a key for a different tenant)
+struct EncryptedRangeFileWriter : public IRangeFileWriter {
+	struct Options {
+		constexpr static FileIdentifier file_identifier = 3152016;
+
+		bool encryptionEnabled;
```

Comment:
nit: shouldn't the class only be instantiated if `encryption` is enabled? Maintaining  a bool flag seems redundant. Instead, we might add an assert to ensure the `encryptionEnabled == true` ?

Reply:
Ah yeah the point of this options struct was to be able to add stuff easily (such as a flag for compression). For now I will change this to `compressionEnabled` but always make it default to `false`

Reply:
I would remove this `bool` as the class name says `EncryptedRangeFileWriter`, if compression needs to be supported we might want to then add a flag to control that behavior, especially because `compression` can be enabled/disabled irrespective of `encryption`